spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+It represents selection using complex placeholders,
+which has the side effect of allowing literal text and other placeholders
+before and after the top-level selection placeholder.
+However, for reliable translations,
+there should be no translatable contents before or after the selection placeholder;
+instead, each selectable pattern should form one complete “translation unit”.
+Because the existing ICU MessageFormat starts in “text mode”,
+even though it looks like there is no extraneous text,
+spurious white space creeps in from developers’ line breaking of long message strings.
+The remedy is to always use syntax to indicate the start of translatable contents.
+
+We use curly braces to delimit patterns because
+`{}` are the paired ASCII punctuation characters least commonly used in normal text.
+For the same reason, we also use them for embedding placeholders in patterns.
+
+Literal text can use any characters except for curly braces,
+and except for the backslash, which we use as usual for escaping.
+That is, the only special characters inside a pattern are `{}\`.
+The only allowed escape sequences are `\{`, `\}`, and `\\`.
+It is an error if `\` is followed by any other character.
+
+The message syntax does not use `'` or `"`,
+so that it is easy to hard-code message strings in programming language source code.
+
+# Placeholders
+
+Formatting a message replaces placeholders with values based on run-time arguments or special functions.
+We also allow for value literals specified inside the placeholder,
+instead of using an argument name;
+and we also allow for invoking functions without using argument names or value literals.
+```
+{$name}
+{$count :number}
+{$fraction :number style=percent minFractions=2}
+{<25> :number}
+{:specialFunction optionKey=optionValue key2=<value with spaces>}
+```
+
+An argument name is a `$` immediately followed by an identifier.
+A message formatting function will typically accept a Map of argument keys to values
+where the keys match argument name identifiers in the patterns of the message.
+
+TODO: For the definition of identifiers we should consult with the Unicode Source Code Working Group.
+
+If the placeholder specifies only an argument name,
+then the formatting function is inferred from the run-time type of the argument value.
+For example, a string value would simply be inserted,
+and a numeric type could be formatted using some kind of default number formatter.
+- TODO: In the registry, specify the default formatters for a small set of value types.
+
+The function is specified via a `:` immediately followed by an identifier.
+If an argument name or a value literal is given,
+then the function is usually a formatter for its expected input types.
+- TODO: There still seems to be discussion about the function prefix character.
+  It could be some other ASCII punctuation, for example `@`.
+- TODO: Functions must be listed in a registry.
```

Comment:
... or installed by the implementation

Probably specify that unrecognized formats are an error or run toString equivalent?

Reply:
Right, “private use” functions need not be in the CLDR registry. I suspect that each organization would have its own registry of some kind, but mostly what this means is that there is documentation for the name, options, and semantics of each function. I don't expect this sort of registry to be parsed by implementations to actually *implement formatters* -- only to do validation and linting. So the formatter implementations are of course implementation-defined.

I think that a message formatting library should by default fail with an error when it does not recognize a function name. That includes functions that are registered, but not supported by a particular implementation.